### DISCUSSION TITLE
MALAMUTE ROM Development: Jacobian Reduction

### DISCUSSION THREAD

**Discussion by raj-anant** [idaholab/moose#24240] [Q&A General] [answered]
Dear MOOSE Team,

We are developing a ROM for MALAMUTE and are exploring different parts of the problem where we can achieve a reduction. In particular, we are interested in achieving reduction at the level of the Jacobian. However, we had a few concerns about the feasibility of the implementation of any such approach. Specifically, we had the following questions:

(i) Is it possible to read in an approximate base Jocobian for the problem through a file or generate it through a script inside MOOSE?
(ii) Is it possible to make small updates to the base Jacobian as the solution evolves using a script inside MOOSE? Essentially, we want to explore bypassing the computation of the Jacobian through AD at each non-linear iteration.
(iii) Is it possible to bypass PetSci for the solution of the non-linear iteration?
(iv) Is there a convenient way for dumping the Jacobian to a file every few time steps/iterations? Write now we are using "-ksp_view_mat" option, but that dumps the Jacobian directly into the text output file serially at every non-linear iteration, which is very difficult to post-process.   
(v) Similarly, is there a way to dump the residuals to a file every few non-linear iterations? We are looking for the residuals at the start of the iteration, not just the final value after convergence.

We would appreciate any help on this.

Thanks,

Anant

**Comment by GiudGiud** [comment]
@cticenhour

**Reply by jiangwen84** [reply]
@GiudGiud This is more like framework/solver level question. @raj-anant worked with me, and I do not have sufficient knowledge to answer some of his question. We hope posting those questions here will provide us better directions to approach next.

**Comment by lindsayad** [comment, marked as answer]
The short answer to all of your questions is sadly no but we welcome contributions.  You *can* write your own `Executioner` and `SolveObject` and `FEProblemBase` derived objects in which you can do almost anything you want. You could override the Jacobian computation routine from the latter for instance.

I'll give a little more detail on a few:

iii) You can write your own custom line search if that's what you mean by bypassing PETSc's nonlinear solution. However, you would still be using the built-in KSP in the PETSc SNES object
iv) You could hack in some condition `MatView` calls within the framework contingent on time step/iteration number etc.
v) similarly as iv)
